### PR TITLE
Add a crate to test the in-browser light client.

### PR DESCRIPTION
**Pull request by expenses**
At the moment, the wasm light client is at a high risk of being broken by incoming commits. This PR adds a simple test that makes sure the light client connects to peers and starts syncing. This requires #4880 to me merged first.

It's pretty messy, but I'm not sure if there's a great way to improve that.

### CODE REVIEW

File: bin/node/browser-testing/src/ws.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+export function transport() {
```
Comment (tomaka) [inline bin/node/browser-testing/src/ws.js]: I don't have any solution off the top of my head, but we should preferably somehow re-use the `ws.js` that has been copy-pasted a couple times.

File: bin/node/browser-testing/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+use wasm_bindgen_test::{wasm_bindgen_test, wasm_bindgen_test_configure};
+use wasm_bindgen_futures::JsFuture;
+use wasm_bindgen::JsValue;
+use wasm_bindgen::prelude::wasm_bindgen;
+use jsonrpc_core::types::{MethodCall, Success, Version, Params, Id};
+use serde::de::DeserializeOwned;
+
+wasm_bindgen_test_configure!(run_in_browser);
+
+#[wasm_bindgen(module = "/src/ws.js")]
```
Comment (tomaka) [inline bin/node/browser-testing/src/lib.rs]: Idea: what about we do it this way in the real browser node code as well?

File: bin/node/browser-testing/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    serde_json::from_str(&value_string).unwrap()
+}
+
+#[wasm_bindgen_test]
+async fn runs() {
+    let mut client = node_cli::start_client(CHAIN_SPEC.into(), "info".into(), transport())
+            .await
+            .unwrap();
+
+    // Give the node a bit of time to connect to peers.
```
Comment (tomaka) [inline bin/node/browser-testing/src/lib.rs]: This seems very flay.
I'd prefer if the code looked like that:

```rust
let master_timeout = Delay::new(Duration::from_secs(45));
loop {
    // Check that timeout hasn't expired.
    assert!((&mut master_timeout).now_or_never().is_none());

    // Let the node do a bit of work.
    Delay::new(Duration::from_secs(5)).await;

    // break from the loop if the tests succeed
}
```

File: bin/node/browser-testing/src/lib.rs (hunk 2 above)
Comment (expenses) [reply bin/node/browser-testing/src/lib.rs]: Yeah, that'd probably be better than copy-pasting it around. I'll look into that in a bit.

Comment (expenses) [reply bin/node/browser-testing/src/lib.rs]: See https://github.com/libp2p/rust-libp2p/pull/1454.

**Comment by TriplEight** [comment]
Firefox will be added to the CI env in https://github.com/paritytech/scripts/pull/152

**Comment by expenses** [comment]
@tomaka Would you like to wait until https://github.com/libp2p/rust-libp2p/pull/1454 is merged and libp2p 0.16.1 is published before merging this? I'm happy either way.

File: bin/node/browser-testing/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    serde_json::from_str(&value_string).unwrap()
+}
+
+#[wasm_bindgen_test]
+async fn runs() {
+    let mut client = node_cli::start_client(CHAIN_SPEC.into(), "info".into(), transport())
+            .await
+            .unwrap();
+
+    let mut master_timeout = Delay::new(Duration::from_secs(45));
```
Comment (tomaka) [inline bin/node/browser-testing/src/lib.rs]:
```suggestion
    let mut test_timeout = Delay::new(Duration::from_secs(45));
```

File: bin/node/browser-testing/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    loop {
+        // Check that timeout hasn't expired.
+        assert!((&mut master_timeout).now_or_never().is_none());
```
Comment (tomaka) [inline bin/node/browser-testing/src/lib.rs]:
```suggestion
        assert!((&mut test_timeout).now_or_never().is_none());
```

**Comment by expenses** [comment]
The test at present is failing, perhaps due to https://github.com/rustwasm/wasm-bindgen/issues/1801. We could either merge the changes now but remove the CI test, running it manually in a browser when needed or wait until the issue is fixed.

**Comment by expenses** [comment]
Ready to go once https://github.com/rustwasm/wasm-bindgen/pull/2036 is merged.

**Comment by TriplEight** [comment]
@expenses the test has failed
<details>

```
   Compiling node-cli v2.0.0-alpha.3 (/builds/bin/node/cli)
warning: unused imports: `FutureExt`, `future::Future`
  --> utils/prometheus/src/lib.rs:17:20
   |
17 | use futures_util::{FutureExt, future::Future};
   |                    ^^^^^^^^^  ^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused imports: `Encoder`, `TextEncoder`
  --> utils/prometheus/src/lib.rs:26:18
   |
26 | use prometheus::{Encoder, TextEncoder, core::Collector};
   |                  ^^^^^^^  ^^^^^^^^^^^

warning: unused import: `std::io`
  --> client/rpc-servers/src/lib.rs:21:5
   |
21 | use std::io;
   |     ^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `log::error`
  --> client/rpc-servers/src/lib.rs:23:5
   |
23 | use log::error;
   |     ^^^^^^^^^^

warning: constant item is never used: `MAX_PAYLOAD`
  --> client/rpc-servers/src/lib.rs:27:1
   |
27 | const MAX_PAYLOAD: usize = 15 * 1024 * 1024;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: constant item is never used: `WS_MAX_CONNECTIONS`
  --> client/rpc-servers/src/lib.rs:30:1
   |
30 | const WS_MAX_CONNECTIONS: usize = 100;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: unused import: `FrozenForDuration`
  --> client/db/src/light.rs:42:31
   |
42 | use crate::{DatabaseSettings, FrozenForDuration};
   |                               ^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused doc comment
  --> client/offchain/src/api/http_dummy.rs:42:3
   |
42 |         /// Because this always returns an error, none of the other methods should ever be called.
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
43 |         Err(())
   |         ------- rustdoc does not generate documentation for expressions
   |
   = note: `#[warn(unused_doc_comments)]` on by default

warning: function is never used: `deadline_to_future`
  --> client/offchain/src/api/timestamp.rs:51:8
   |
51 | pub fn deadline_to_future(
   |        ^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: unused import: `io`
  --> client/service/src/lib.rs:31:24
   |
31 | use std::{borrow::Cow, io, pin::Pin};
   |                        ^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `std::net::SocketAddr`
  --> client/service/src/lib.rs:33:5
   |
33 | use std::net::SocketAddr;
   |     ^^^^^^^^^^^^^^^^^^^^

warning: unused import: `parity_util_mem::MallocSizeOf`
  --> client/service/src/lib.rs:53:5
   |
53 | use parity_util_mem::MallocSizeOf;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: missing documentation for a trait
  --> client/service/src/lib.rs:81:1
   |
81 | pub trait MallocSizeOfWasm {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> client/service/src/lib.rs:20:9
   |
20 | #![warn(missing_docs)]
   |         ^^^^^^^^^^^^

warning: unused variable: `pool`
  --> client/informant/src/lib.rs:38:6
   |
38 |     let pool = service.transaction_pool();
   |         ^^^^ help: consider prefixing with an underscore: `_pool`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused import: `sc_service::Configuration`
  --> bin/node/cli/src/browser.rs:20:5
   |
20 | use sc_service::Configuration;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: type alias is never used: `ConcreteBlock`
   --> bin/node/cli/src/service.rs:258:1
    |
258 | type ConcreteBlock = node_primitives::Block;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: type alias is never used: `ConcreteClient`
   --> bin/node/cli/src/service.rs:259:1
    |
259 | / type ConcreteClient =
260 | |     Client<
261 | |         Backend<ConcreteBlock>,
262 | |         LocalCallExecutor<Backend<ConcreteBlock>, NativeExecutor<node_executor::Executor>>,
263 | |         ConcreteBlock,
264 | |         node_runtime::RuntimeApi
265 | |     >;
    | |______^

warning: type alias is never used: `ConcreteBackend`
   --> bin/node/cli/src/service.rs:266:1
    |
266 | type ConcreteBackend = Backend<ConcreteBlock>;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: type alias is never used: `ConcreteTransactionPool`
   --> bin/node/cli/src/service.rs:267:1
    |
267 | / type ConcreteTransactionPool = sc_transaction_pool::BasicPool<
268 | |     sc_transaction_pool::FullChainApi<ConcreteClient, ConcreteBlock>,
269 | |     ConcreteBlock
270 | | >;
    | |__^

warning: function is never used: `new_full`
   --> bin/node/cli/src/service.rs:276:8
    |
276 | pub fn new_full(config: NodeConfiguration)
    |        ^^^^^^^^

warning: method is never used: `load`
  --> bin/node/cli/src/lib.rs:66:2
   |
66 |     pub(crate) fn load(self) -> Result<chain_spec::ChainSpec, String> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: method is never used: `from`
  --> bin/node/cli/src/lib.rs:75:2
   |
75 |     pub(crate) fn from(s: &str) -> Option<Self> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: function is never used: `load_spec`
  --> bin/node/cli/src/lib.rs:86:4
   |
86 | fn load_spec(id: &str) -> Result<Option<chain_spec::ChainSpec>, String> {
   |    ^^^^^^^^^

^[[A^[[A   Compiling node-browser-testing v2.0.0 (/builds/bin/node/browser-testing)
warning: unused import: `wasm_bindgen::prelude::wasm_bindgen`
  --> bin/node/browser-testing/src/lib.rs:33:5
   |
33 | use wasm_bindgen::prelude::wasm_bindgen;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `wasm_bindgen::prelude::wasm_bindgen`
  --> bin/node/browser-testing/src/lib.rs:33:5
   |
33 | use wasm_bindgen::prelude::wasm_bindgen;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default
    Finished test [unoptimized + debuginfo] target(s) in 28.12s
     Running /ci-cache/targets/master/testjob/wasm32-unknown-unknown/debug/deps/node_browser_testing-3bc0806d52c96dfa.wasm
/ci-cache/targets/master/testjob/wasm32-unknown-unknown/debug/deps/node_browser_testing-3bc0806d52c96dfa.wasm: 1: /ci-cache/targets/master/testjob/wasm32-unknown-unknown/debug/deps/node_browser_testing-3bc0806d52c96dfa.wasm: Syntax error: end of file unexpected
error: test failed, to rerun pass '-p node-browser-testing --lib'
```

</details>

**Comment by bkchr** [comment]
@expenses status?

**Comment by expenses** [comment]
@bkchr waiting on a new release of [wasm-bindgen](https://github.com/rustwasm/wasm-bindgen) that contains https://github.com/rustwasm/wasm-bindgen/pull/2036.

**Comment by tomaka** [comment]
@expenses It's been published https://github.com/rustwasm/wasm-bindgen/pull/2051 and Substrate has apparently already been updated.

**Comment by bkchr** [comment]
@expenses please merge master and please fix the new warnings.

**Comment by expenses** [comment]
Hey @TriplEight, I've tried all I could, but I'm really struggling the get the test to work properly on firefox, so I've set it up to use chromedriver instead, which it does work on. Do you think you could please install chrome/chromium and chromedriver with matching versions on the build host? Thanks!

**Comment by expenses** [comment]
I've merged in https://github.com/paritytech/substrate/pull/5613.

**Comment by bkchr** [comment]
Still not working?

**Comment by expenses** [comment]
> Still not working?

It worked once: https://gitlab.parity.io/parity/substrate/-/jobs/459858, but then when i re-ran that test it failed, with the only different being the port number that `wasm-pack-test-runner` randomly selects (I haven't noticed a pattern about which ports work and which don't though).

**Comment by bkchr** [comment]
Hmm :(

**Comment by bkchr** [comment]
@expenses Did you tried running it with `verbose` option enabled to better debug this?

**Comment by gavofyork** [comment]
@expenses ?

**Comment by TriplEight** [comment]
currently, it doesn't work with the chromedriver, working on it.

**Comment by TriplEight** [comment]
@expenses 
tried with your new branch  https://github.com/expenses/wasm-bindgen/tree/port-env-var with `WASM_BINDGEN_TEST_ADDRESS` set to `127.0.0.1:37351` (also with 9515). 
It's reproducible with 
```
docker run -it -v ~/cache/substrate:/ci-cache parity/rust-builder:latest
# and there 
export SCCACHE_DIR=/ci-cache/sccache
export CARGO_HOME=/ci-cache/cargo
export CARGO_TARGET_DIR=/ci-cache/targets/
export RUSTFLAGS="-Cdebug-assertions=y"
git clone https://github.com/paritytech/substrate .
git checkout ashley-browser-testing
export CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER="wasm-bindgen-test-runner"
export WASM_BINDGEN_TEST_TIMEOUT=60
cargo +nightly test --target wasm32-unknown-unknown -p node-browser-testing -Z features=itarget
```
and
webdriver.json
```
{"goog:chromeOptions": {
    "args": [
      "--verbose",
      "whitelisted-ips=127.0.0.1",
      "--headless",
      "--disable-gpu",
      "--no-sandbox"
    ]
  }
}
```

It did not work:
<details>
<summary>logs</summary>

```
    Finished test [unoptimized + debuginfo] target(s) in 0.47s
     Running /ci-cache/targets/wasm32-unknown-unknown/debug/deps/node_browser_testing-3ef5d09ca18e450f.wasm
Set timeout to 60 seconds...
Running headless tests in Chrome on `http://127.0.0.1:9515/`
Try find `webdriver.json` for configure browser's capabilities:
Ok
Failed to detect test as having been run. It might have timed out.
output div contained:
    running 1 test

driver status: signal: 9                          
driver stdout:
    Starting ChromeDriver 80.0.3987.162 (f2c5dd6138153bb0658091205bd1a1717f16081a-refs/branch-heads/3987@{#1034}) on port 9515
    Only local connections are allowed.
    Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.

driver stderr:
    [1587141083.320][SEVERE]: bind() failed: Cannot assign requested address (99)

Error: some tests failed                          
error: test failed, to rerun pass '-p node-browser-testing --lib'

# 
root@83b8c00b60ca:/builds# chromedriver --whitelisted-ips=127.0.0.1 --verbose
Starting ChromeDriver 80.0.3987.162 (f2c5dd6138153bb0658091205bd1a1717f16081a-refs/branch-heads/3987@{#1034}) on port 9515
Remote connections are allowed by a whitelist (127.0.0.1).
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
```
</details>

Apparently, this setup does not read `webdriver.json` even though it tells that finds it:

```
Try find `webdriver.json` for configure browser's capabilities:
Ok
```

Proof: 

in `cargo test` run there's `Only local connections are allowed.` and if run `chromedriver --whitelisted-ips=127.0.0.1` it runs with `Remote connections are allowed by a whitelist (127.0.0.1).` 
Native `chromedriver`'s port is `9515` 
I have no clue how to pass flags to it, so I stuck with debug here.

**Comment by expenses** [comment]
@TriplEight Oh, thanks for figuring out how to make it reproducible!

**Comment by expenses** [comment]
I think I've finally gotten it to work 😊😊😊😊

**Pull request merged**